### PR TITLE
Endpoints for comments Misha

### PR DESCRIPTION
**Pull request by DoctorHeisenberg**
Step to steps:
1. Start the app, open Swagger
2. Register or log in
3. Make sure that you contain in workspace
4. Create new, or choose the existing task
5. Now you can:
     I) See all task comments
     II) Write comments
     III) Edit comments
     IV) Delete your own comments, or somebody's (if you are owner of workspace)

### CODE REVIEW

File: Provis.Core/DTO/CommentDTO/CommentListDTO.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+
+namespace Provis.Core.DTO.CommentDTO
+{
+    public class CommentListDTO
+    {
+        public int Id { get; set; }
+        public string CommentText { get; set; }
+        public DateTime dateTime { get; set; }
```
Comment (AndriyKovalets) [inline Provis.Core/DTO/CommentDTO/CommentListDTO.cs]: Change naming in 9 line, please

File: Provis.Core/Interfaces/Services/ICommentService.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using Provis.Core.DTO.CommentDTO;
+using Provis.Core.DTO.CommentsDTO;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Provis.Core.Interfaces.Services
+{
+    public interface ICommentService
+    {
+        Task<List<CommentListDTO>> GetCommentsAsync(int taskId);
```
Comment (AndriyKovalets) [inline Provis.Core/Interfaces/Services/ICommentService.cs]: Change name from "GetCommentsAsync" on "GetCommentListAsync"

File: Provis.Core/Interfaces/Services/ICommentService.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        Task CommentAsync(CreateCommentDTO commentDTO, string userId);
```
Comment (AndriyKovalets) [inline Provis.Core/Interfaces/Services/ICommentService.cs]: I think that "CreateCommentAsync" or "AddCommentAsync" is more informative than "CommentAsync"

File: Provis.Core/Services/CommentService.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+        public async Task CommentAsync(CreateCommentDTO commentDTO, string userId)
+        {
+            Comment comment = new()
+            {
+                CommentText = commentDTO.CommentText,
+                DateOfCreate = DateTime.UtcNow,
+                TaskId = commentDTO.TaskId,
+                UserId = userId
+            };
```
Comment (AndriyKovalets) [inline Provis.Core/Services/CommentService.cs]: Use mapper for it.

File: Provis.Core/Services/CommentService.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            await _commentRepository.AddAsync(comment);
+            await _commentRepository.SaveChangesAsync();
+        }
+
+        public async Task<List<CommentListDTO>> GetCommentsAsync(int taskId)
+        {
+            var specification = new Comments.CommentTask(taskId);
+            var commentList = await _commentRepository.GetListBySpecAsync(specification);
+
+            var listCommentsToReturn = _mapper.Map<List<CommentListDTO>>(commentList);
```
Comment (AndriyKovalets) [inline Provis.Core/Services/CommentService.cs]: Change name from "listCommentsToReturn" to "commentListToReturn"

File: Provis.Core/Services/CommentService.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            var userWorkspace = await _userWorkspaceRepository.GetFirstBySpecAsync(specification);
+
+            var comment = await _commentRepository.GetByKeyAsync(id);
+
+            if (!(comment.UserId == userId || 
+                userWorkspace.RoleId == (int)WorkSpaceRoles.OwnerId))
+            {
+                throw new HttpException(System.Net.HttpStatusCode.Forbidden,
+                    "Only cretor or workspace owner can delete comments");
+            }
```
Comment (AndriyKovalets) [inline Provis.Core/Services/CommentService.cs:99]: In requirements is written that task creator can delete comment too.

File: Provis.WebApi/Controllers/CommentController.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        [Authorize]
+        [HttpPost]
+        [WorkspaceRoles(new WorkSpaceRoles[] { 
+            WorkSpaceRoles.OwnerId,
+            WorkSpaceRoles.ManagerId,
+            WorkSpaceRoles.MemberId,
+            WorkSpaceRoles.ViewerId})]
+        [Route("comment")]
```
Comment (AndriyKovalets) [inline Provis.WebApi/Controllers/CommentController.cs]: there needn't write "comment" in route.

File: Provis.WebApi/Controllers/CommentController.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+        [Authorize]
+        [HttpGet]
+        [WorkspaceRoles(new WorkSpaceRoles[] {
+            WorkSpaceRoles.OwnerId,
+            WorkSpaceRoles.ManagerId,
+            WorkSpaceRoles.MemberId,
+            WorkSpaceRoles.ViewerId})]
+        [Route("list")]
+        public async Task<IActionResult> GetComments(int taskId)
```
Comment (AndriyKovalets) [inline Provis.WebApi/Controllers/CommentController.cs]: Change name form "GetComments" to "GetCommentListAsync"

File: Provis.WebApi/Controllers/CommentController.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        [Authorize]
+        [HttpPut]
+        [WorkspaceRoles(new WorkSpaceRoles[] {
+            WorkSpaceRoles.OwnerId,
+            WorkSpaceRoles.ManagerId,
+            WorkSpaceRoles.MemberId,
+            WorkSpaceRoles.ViewerId})]
+        [Route("edit")]
```
Comment (AndriyKovalets) [inline Provis.WebApi/Controllers/CommentController.cs]: there needn't write "edit" in route becase you use put http method.

File: Provis.Core/DTO/CommentDTO/CreateCommentDTO.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Provis.Core.DTO.CommentsDTO
+{
+    public class CreateCommentDTO
+    {
+        public string CommentText { get; set; }
+        public int TaskId { get; set; }
+        public int WorkspaceId { get; set; }
+    }
```
Comment (AndriyKovalets) [inline Provis.Core/DTO/CommentDTO/CreateCommentDTO.cs:10]: why there is no validator for it.

File: Provis.Core/DTO/CommentDTO/EditCommentDTO.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Provis.Core.DTO.CommentsDTO
+{
+    public class EditCommentDTO
+    {
+        public int CommentId { get; set; }
+        public string CommentText { get; set; }
+        public int WorkspaceId { get; set; }
```
Comment (AndriyKovalets) [inline Provis.Core/DTO/CommentDTO/EditCommentDTO.cs:9]: why there is no validator for it.

**Review by vladtymo — APPROVED** [review]

**Review by AndriyKovalets — APPROVED** [review]

**Pull request merged**